Update R2R format doc to 3.0

trylek opened the pull request:
I have updated the various enumerations and tables. I didn't try
to go overboard w.r.t. understanding tables I'm not yet familiar
with, most notably the diagnostic, inlining and profiling tables.

Thanks

Tomas

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -225,22 +236,67 @@ additional data determined by the flags.
 image with pointers to their unwind info. The standard Windows xdata/pdata format is used.
 ARM format is used for x86 to compensate for lack of x86 unwind info standard.
-The unwind info blob is immediately followed by gc info blob.
+The unwind info blob is immediately followed by GC info blob. The encoding slightly differs for amd64 which encodes an extra 4-byte  representing the end RVA of the frame info blob.
```
```suggestion
The unwind info blob is immediately followed by GC info blob. The encoding slightly differs for amd64 which encodes an extra 4-byte representing the end RVA of the frame info blob.
```

trylek replied:
Fixed in 2nd commit, thanks for spotting.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -225,22 +236,67 @@ additional data determined by the flags.
+
+| Offset in auxiliary data | Size | Content |
+|-------------------------:|-----:|:--------|
+| 0 | 4 | Offset to GC ref map info for method #0 relative to this byte i.e. 4 * (MethodCount / 1024 + 1) |
```
The markdown should be pleasant to read as raw text. Could you please add spaces to these tables as necessary?

trylek replied:
Does this also mean we should use line breaks within paragraphs to "manually word-wrap"? I looked at various CoreCLR documents and they usually didn't; after all we probably cannot do that within tables.

trylek replied:
For now (as of 2nd commit) I have manually aligned the text tables except for the variable-length last column.

jkotas replied:
>  except for the variable-length last column.

It does not look very pleasant. I would align the last column too. E.g. Would you format the code comment this way?

trylek replied:
Based on your suggestion I have aligned all tables in 5th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -225,22 +236,67 @@ additional data determined by the flags.
+
+| CORCOMPILE_GCREFMAP_TOKENS | Value | Stack frame entry interpretation |
+|:---------------------------|------:|:--------|
+| ``GCREFMAP_SKIP`` | 0 | Not a GC-relevant entry |
```
Is there a good reason to use double ticks for escaping? I would think a single tick should be enough. (This was in the original doc - but I have not noticed it.)

trylek replied:
I was under the impression that the double tick is supposed to be something like a "code quote" but perhaps I was mistaken. I started to learn md yesterday.

trylek replied:
In fact, in the GitHub PR rendering the "code quotes" look pretty annoying. I'm going to remove them everywhere.

trylek replied:
Fixed in 2nd commit (in tables anyway; in normal text they look good to me).

jkotas replied:
The common markdown convention is to use single ticks to quote identifiers in the text and triple ticks to quote code fragments. I have looked at the other docs inside botr directory and all of them use single and triple ticks. I am not able to find any other instance of double ticks.

I think if you replace "``" with "`", everything is still going to render the same.

trylek replied:
Globally replaced in 10th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -114,6 +112,13 @@ enum ReadyToRunSectionType
     READYTORUN_SECTION_EXCEPTION_INFO               = 104,
```
Typo ~10 lines above: `EADYTORUN_SECTION`

trylek replied:
Thanks for spotting, fixed in 2nd commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -225,22 +236,67 @@ additional data determined by the flags.
+| ``GCREFMAP_TYPE_PARAM`` | 4 | Hidden type instantiation argument to generic method |
+| ``GCREFMAP_VASIG_COOKIE`` | 5 | VARARG signature cookie |
+
+The position values are calculated in ``size_t`` aka ``IntPtr`` units (4 bytes for 32-bit architectures vs. 8 bytes for 64-bit architectures) and start at the offset ``TransitionBlock::GetOffsetOfFirstGCRefMapSlot`` relative to the transition frame address. For x86 the position encoding is [somewhat more complicated](https://github.com/dotnet/coreclr/blob/d5d18896900561b7aaf38ba9501a8525a4b9caea/ ...
```
This should not be referencing the implementation. 

The position is a index into a logical array that starts with spilled argument registers, and continues with stack registers. This is true for both x86 and non-x86, I think.

trylek replied:
Well, it's just that on x86 the spilled argument registers are indexed in reverse order (or perhaps they are somehow spilled in reverse order, I actually don't know what's the exact problem there). I can easily remove the reference to the implementation if you find it inappropriate though.

trylek replied:
Not addressed in 2nd commit - waiting on your further feedback.

jkotas replied:
>  I can easily remove the reference to the implementation if you find it inappropriate though.

Yes, I think it would be best to not reference the implementation here.

trylek replied:
Removed in 3rd commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -225,22 +236,67 @@ additional data determined by the flags.
+|-------:|-----:|:------|
+| 0 | 4 | Frame info start RVA |
+| 4 | 4 | Frame info end RVA (1 plus RVA of last byte) |
+| 8 | 4 | GC info start RVA |
```
Technically, it is Unwind info start...

trylek replied:
Fixed in 2nd commit, thanks for keeping an eye on proper terminology, I never get that right.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -326,6 +382,47 @@ This section marks region that contains thunks for ``READYTORUN_HELPER_DelayLoad
 
+## READYTORUN_SECTION_AVAILABLE_TYPES
+
+This section contains a native hashtable of all defined & export types within the compilation module. The key is a checksum of the full type name, the value is the exported type or defined type token row ID left-shifted by one and or-ed with bit 0 defining the token type:
```
The key is the full type name.

trylek replied:
Hmm, I don't understand, can you please clarify? According to what I see in

https://github.com/dotnet/coreclr/blob/93675fbf467f54ab0f1f5d183c70750c9822c9ca/src/zap/zapreadytorun.cpp#L484

the key is the hash (should I have said 'hash' instead of 'checksum'?) of the full type name, not the type name itself.

trylek replied:
Renamed "checksum" to "hash" / "hashcode". Didn't change "hash of full type name" to "full type name" as I believe the vertex hashtable has the hash, not the full type name, as the key, but I'm prepared to be proven wrong and fix my mistake.

jkotas replied:
https://en.wikipedia.org/wiki/Hash_table

The vertex table does not contain copy of the key inline; but it does not change the fact the key is the full type name.

trylek replied:
OK, I buy that, will fix in 3rd commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -326,6 +382,47 @@ This section marks region that contains thunks for ``READYTORUN_HELPER_DelayLoad
+| 0         | defined type |
+| 1         | exported type |
+
+The checksum algorithm is implemented in [vm/versionresilienthashcode.cpp](https://github.com/dotnet/coreclr/blob/ec2a74e7649f1c0ecff32ce86724bf3ca80bfd46/src/vm/versionresilienthashcode.cpp#L75).
```
Checksums are used to detect data errors. This is hashcode, not checksum.

trylek replied:
OK, so it's just about the terminology thing checksum -> hash? Thanks for pointing that out, I'll fix that.

trylek replied:
Corrected in 2nd commit.

jkotas replied:
Right, https://en.wikipedia.org/wiki/Checksum

trylek commented:
@jkotas I have addressed most of your PR feedback in 2nd commit except for two items - I'm not yet fully clear on the x86 argument spilling order thing and and on whether it's the actual full type name or its hash that forms the key of the AVAILABLE_TYPES vertex hashtable.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -221,26 +232,71 @@ additional data determined by the flags.
+| GCREFMAP_TYPE_PARAM        |     4 | Hidden type instantiation argument to generic method |
+| GCREFMAP_VASIG_COOKIE      |     5 | VARARG signature cookie |
+
+The position values are calculated in ``size_t`` aka ``IntPtr`` units (4 bytes for 32-bit architectures vs. 8 bytes for 64-bit architectures) and start at the offset ``TransitionBlock::GetOffsetOfFirstGCRefMapSlot`` relative to the transition frame address.
```
It would be nice to explain this without referencing `TransitionBlock::GetOffsetOfFirstGCRefMapSlot` at all.

trylek replied:
Reformulated in 5th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -221,26 +232,71 @@ additional data determined by the flags.
-compressed argument maps that allow precise GC stack scanning while the helper is running.
+compressed argument maps that allow precise GC stack scanning while the helper is running. The CoreCLR runtime class [``GCRefMapDecoder``](https://github.com/dotnet/coreclr/blob/6b9a3d3a87825b1a34bd8f114c9b181ce75b3b2e/src/inc/gcrefmap.h#L157) is used to parse this information. This data would not be required for runtimes that allow conservative stack scanning.
+
+The auxiliary data table is supposed to contain the exact same number of GC ref map records as there are method entries in the import section. To accelerate GC ref map lookup, the auxiliary data section starts with a lookup table holding the offset of every 1024-th method in the runtime function table within the linearized GC ref map.
```
```suggestion
The auxiliary data table contains the exact same number of GC ref map records as there are method entries in the import section. To accelerate GC ref map lookup, the auxiliary data section starts with a lookup table holding the offset of every 1024-th method in the runtime function table within the linearized GC ref map.
```

trylek replied:
Fixed in 5th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -483,9 +597,12 @@ enum ReadyToRunHelper
     READYTORUN_HELPER_EndCatch                  = 0x110,
+
+    // A flag to indicate that a helper call uses VSD
+    READYTORUN_HELPER_FLAG_VSD                  = 0x10000000,
```
This internal crossgen implementation hack. It is not part of the file format spec. Correct? This document should be the file format spec, insulated from the implementation hacks.

trylek replied:
OK, will remove that. In fact I originally ported this to CPAOT as I was under the impression that it actually is a part of the format but upon closer scrutiny I concur it's just an internal contract within zapinfo. I'll take a look whether I can easily remove it from CPAOT too.

trylek replied:
Removed in 6th commit.

jkotas replied:
I would be nice to delete it from CPAOT and R2R dump too.

trylek replied:
Sure, will do.

trylek commented:
@jkotas - I believe I have addressed all your feedback so far. Could you please take another look when you have a chance?

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -326,6 +382,49 @@ This section marks region that contains thunks for ``READYTORUN_HELPER_DelayLoad
+
+## READYTORUN_SECTION_MANIFEST_METADATA
+
+Manifest metadata is an [ECMA-335] metadata blob containing extra reference assemblies introduced by inlining on top of assembly references stored in the input MSIL. This can only happen for non-versionable methods or for methods within the same *version bubble*.
```
Do we actually use this for non-versionable methods? I do not think that this blob as designed would be sufficient for non-versionable methods outside current version bubble.

trylek replied:
Hmm, you're right this merits further clarification. I meant to say that inlining beyond the version bubble boundary only happens in the presence of non-versionable methods but it's true that we cannot use the manifest metadata for expressing tokens used by the non-versionable method from outside of the version bubble as we cannot rely on row ID's in external assemblies beyond the version bubble boundary to remain constant.

trylek replied:
[Hopefully] improved in 9th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -326,6 +382,49 @@ This section marks region that contains thunks for ``READYTORUN_HELPER_DelayLoad
+| 1 <= *i* <= **ILAR**        | *i* is the index into the MSIL `AssemblyRef` table                                               |
+| *i* > **ILAR**              | *i* - **ILAR** - 1 is the zero-based index into the `AssemblyRef` table in the manifest metadata |
+
+**Note:** This means that the entry corresponding to *i* = **ILAR** + 1 is actually undefined as it corresponds to the `NULL` entry (ROWID #0) in the manifest metadata AssemblyRef table. The first meaningful index into the manifest metadata, *i* = **ILAR** + 2, corresponding to ROWID #1, is historically filled in by Crossgen with the input assembly info but this shouldn't be depended upon, in fac ...
```
The crossgen implementation quirks should not be part of the file format format spec.

trylek replied:
Hmm, I would rather say that this is a runtime contract regarding how the external module addressing is designed. There's no way to encode non-null row #0 in the manifest metadata as the ECMA335 metadata format simply doesn't support that. So the runtime will always end up with a null row when querying ILAR + 1 no matter whether the executable was produced by Crossgen, CPAOT or any other compiler conforming to the current format design. Could you please suggest what you would perceive as an appropriate way to treat this in the format doc?

jkotas replied:
We just took what happened to be there for fragile NGen. The implementation does not quite make sense as many things in fragile NGen.

We are not actually shipping this as officially supported because of the implementation is incomplete. Maybe the manifest metadata should be marked as work-in-process on future version here. We may want to change and clean this up so that it make sense.

trylek replied:
I have added a disclaimer echoing your feedback to the doc in 10th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -326,6 +382,49 @@ This section marks region that contains thunks for ``READYTORUN_HELPER_DelayLoad
+
+Manifest metadata is an [ECMA-335] metadata blob containing extra reference assemblies introduced by inlining on top of assembly references stored in the input MSIL. This can only happen for non-versionable methods or for methods within the same *version bubble*.
+
+As of R2R version 3.0, The metadata is only searched for the AssemblyRef table. This is used to translate module override indices in signatures to the actual reference modules (using either the ``READYTORUN_FIXUP_ModuleOverride`` bit flag on the signature fixup byte or the ``ELEMENT_TYPE_MODULE_ZAPSIG`` COR element type).
```
```suggestion
As of R2R version 3.0, the metadata is only searched for the AssemblyRef table. This is used to translate module override indices in signatures to the actual reference modules (using either the ``READYTORUN_FIXUP_ModuleOverride`` bit flag on the signature fixup byte or the ``ELEMENT_TYPE_MODULE_ZAPSIG`` COR element type).
```

trylek replied:
Thanks for spotting the typo, fixed in 8th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -1,15 +1,17 @@
 # PE Headers and CLI Headers
 
-ReadyToRun v1 images conform to CLI file format as described in ECMA-335, with the following 
+ReadyToRun v1-3 images conform to CLI file format as described in ECMA-335, with the following 
```
```suggestion
ReadyToRun images conform to CLI file format as described in ECMA-335, with the following 
```

trylek replied:
Removed in 8th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -70,7 +68,7 @@ ReadyToRun images from other CLI images with ManagedNativeHeader (e.g. NGen imag
 ### READYTORUN_HEADER::MajorVersion/MinorVersion
 
-The current format version is 1.1. MajorVersion increments are meant for file format breaking changes. 
+The current format version is 3.0. MajorVersion increments are meant for file format breaking changes. 
```
```suggestion
The current format version is 3.1. MajorVersion increments are meant for file format breaking changes. 
```

The tables below talk about 3.1

trylek replied:
Thanks for pointing that out, fixed in 8th commit.

jkotas commented on Documentation/botr/readytorun-format.md:
```diff
@@ -79,9 +77,9 @@ native code from image of version 3.0.
-|:-----|------:|:------------|
-| ``READYTORUN_FLAG_PLATFORM_NEUTRAL_SOURCE`` | 0x00000001 | Set if the original IL image was platform neutral. The platform neutrality is part of assembly name. This flag can be used to reconstruct the full original assembly name. |
+| Flag                                    |      Value | Description                                                                                                                                                                |
+|:----------------------------------------|-----------:|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
```
Hmm, these long lines do not make it much readable as a raw text in simple editor either. I guess it may better to leave the tables with long multi-line cells without any indentation. I am sorry for the back-and-forth.

trylek replied:
No worries, I myself am learning the markdown syntax as I go. One compromise option that occurs to me is as follows: I found out in the documentation that the outer vertical bars are optional. So I'd just leave them out and I'd shorten the minuses to just cover the header row. Apart from being visually somewhat closer to the rich-formatted md tables, I believe it would make the tables more readable in text mode too. I'll try that out in 8th commit and if you don't like it, I'll revert it back to the original version.

trylek replied:
OK, slight correction. I had to keep the initial vertical bar as otherwise the parser doesn't recognize the table if the leftmost column is right-aligned (typically for numeric values) and I believe that artificially left-aligning them just to satisfy the parser is not visually pleasing in the text view. For now I'm relatively content that the updated version in 8th commit looks much better than all the previous versions.

jkotas replied:
Agree. It looks pretty good now.

trylek commented:
@jkotas - OK, I think I'm mostly done with your last batch of comments modulo my uncertainty w.r.t. desired level of detail in the manifest metadata addressing calculation description. I have applied all the mechanical changes you suggested. I tried to improve the introductory paragraph on manifest metadata and I removed the final vertical bars from tables - I believe it makes them much prettier in text mode. Please take another look when you have a chance.

jkotas commented:
Looks good to me otherwise. Thank you!
